src/main/java/seedu/mtracker/filemanager/CryptoDecoder.java

```diff
@@ -1,25 +1,82 @@
+import seedu.mtracker.error.fileerror.InvalidRemarksInFileError;
 import seedu.mtracker.model.Instrument;
 import seedu.mtracker.model.InstrumentManager;
 import seedu.mtracker.model.subinstrument.Crypto;
+import seedu.mtracker.ui.TextUi;
 
 import java.time.LocalDate;
 
 public class CryptoDecoder extends InstrumentDecoder {
 
     public static final int CRYPTO_EXPIRY_INDEX = 5;
     public static final int CRYPTO_REMARKS_INDEX = 6;
+    protected static LocalDate decodedExpiry;
+    protected static String decodedRemarks;
+
+    public static boolean isValidExpiry(String[] textSegment) {
+        return Validate.isValidExpiry(textSegment[CRYPTO_EXPIRY_INDEX]);
+    }
+
+    public static String getExpiryFromFile(String[] textSegment) throws InvalidEmptyExpiryInFileError {
+        String expiry;
+        try {
+            expiry = textSegment[CRYPTO_EXPIRY_INDEX];
+        } catch (IllegalArgumentException e) {
+            throw new InvalidEmptyExpiryInFileError();
+        }
+        return expiry;
+    }
+
+    public static String getRemarksFromFile(String[] textSegment) throws InvalidRemarksInFileError {
+        String remarks;
+        try {
+            remarks = textSegment[CRYPTO_REMARKS_INDEX];
+        } catch (IllegalArgumentException e) {
+            throw new InvalidRemarksInFileError();
+        }
+        return remarks;
+    }
+
+    public static void validateAndDecodeSpecificAttributes(String[] textSegment) throws InvalidEmptyExpiryInFileError,
+            InvalidRemarksInFileError, InvalidExpirySavedInFileError {
+        String expiry = getExpiryFromFile(textSegment);
+        String remarks = getRemarksFromFile(textSegment);
+        if (!isValidExpiry(textSegment)) {
+            throw new InvalidExpirySavedInFileError();
+        }
+        decodeSpecificAttributes(expiry, remarks);
+    }
+
+    public static void tryValidateAndDecodeSpecificAttributes(String[] textSegment) {
+        try {
+            validateAndDecodeSpecificAttributes(textSegment);
+        } catch (Exception e) {
+            TextUi.showErrorMessage(e);
+        }
+    }
+
+    private static void decodeSpecificAttributes(String expiry, String remarks) {
+        decodedExpiry = LocalDate.parse(expiry);
+        decodedRemarks = remarks;
+    }
 
     public static void addCryptoToList(String[] textSegment, InstrumentManager instrumentManager) {
-        decodeGeneralAttributes(textSegment);
-        // todo: abstract out to decode instrument attributes for the different decoders
-        LocalDate decodedExpiry = LocalDate.parse(textSegment[CRYPTO_EXPIRY_INDEX]);
-        String decodedRemarks = textSegment[CRYPTO_REMARKS_INDEX];
-        Instrument crypto = new Crypto(decodedName, decodedCurrPrice, decodedSentiment,
-                decodedExpiry, decodedRemarks);
+        tryValidateAndDecodeGeneralAttributes(textSegment);
+        tryValidateAndDecodeSpecificAttributes(textSegment);
```

Comment:
To fix the error I think you just have to remove this method and use validateAndDecodeSpecificAttributes(textSegment);
instead. The error handling shld be done to the loop that is calling this